buildbot.py

```diff
@@ -118,15 +123,69 @@ def exec_make(target):
     print(' '.join(args))
     sys.exit(subprocess.call(args))
 
+# mspdbsrv is the service used by Visual Studio to collect debug
+# data during compilation.  One instance is shared by all C++
+# compiler instances and threads.  It poses a unique challenge in
+# several ways:
+#
+# - If not running when the build job starts, the build job will
+#   automatically spawn it as soon as it needs to emit debug symbols.
+#   There's no way to prevent this from happening.
+#
+# - The build job _doesn't_ automatically clean it up when it finishes
+#
+# - By default, mspdbsrv inherits its parent process' file handles,
+#   including (unfortunately) some log handles owned by Buildbot.  This
+#   can prevent Buildbot from detecting that the compile job is finished
+#
+# - If a compile job starts and detects an instance of mspdbsrv already
+#   running, by default it will reuse it.  So, if you have a compile
+#   job A running, and start a second job B, job B will use job A's
+#   instance of mspdbsrv.  If you kill mspdbsrv when job A finishes,
+#   job B will die horribly.  To make matters worse, the version of
+#   mspdbsrv should match the version of Visual Studio being used.
+#
+# This class works around these problems:
+#
+# - It sets the _MSPDBSRV_ENDPOINT_ to a value that's probably unique to
+#   the build, to prevent other builds on the same machine from sharing
+#   the same mspdbsrv endpoint
+#
+# - It launches mspdbsrv with _all_ file handles closed, so that it
+#   can't block the build from being detected as finished.
+#
+# - It explicitly kills mspdbsrv after the build job has finished.
+#
+# - It wraps all of this into a context manager, so mspdbsrv gets killed
+#   even if a Python exception causes a non-local exit.
+class UniqueMspdbsrv(object):
+    def __enter__(self):
+        os.environ['_MSPDBSRV_ENDPOINT_'] = '{}-{}-{}'.format(
```

Comment:
Would adding a -<uuid> to this string help ensure complete uniqueness? Just in case the mspdbsrv does not quit from a previous identical run, it would mean a subsequent identical build would use a distinct instance.